src/decode_all.rs

```diff
@@ -0,0 +1,108 @@
+// Copyright 2017, 2018 Parity Technologies
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{Error, Decode};
+
+/// The error message returned when `decode_all` fails.
+const DECODE_ALL_ERR_MSG: &str = "Input buffer has still data left after decoding!";
+
+/// Extension trait to [`Decode`] that ensures that the given input data is consumed completly while
+/// decoding.
+pub trait DecodeAll: Sized {
+	/// Decode `Self` and consume all of the given input data. If not all data is consumed, an error
+	/// is returned.
+	fn decode_all(input: &[u8]) -> Result<Self, Error>;
```

Comment:
At some point I feel like it would be better to have an extension of the signature as well instead of a restriction:

We can create a trait 
```rust
TerminableInput: Input {
  fn terminated(&mut self) -> bool;
}
```
and having decode_all working on this input

Reply:
Hmm, yeah maybe.

Reply:
But maybe we can come up with a better name?

Reply:
InputEnd, or InputEndable with a is_ended or ended method

Reply:
The thing is that almost every input is endable. Maybe we should just support any input and add the `is_eof` method to input.

Reply:
yes it is a possibility but at the same time is_eof seems to be never needed for codec because of its design. Maybe someone could want not to have to implement this method, but I can't really get an example

Reply:
I thought that you never want to have decodeall on something different than a slice of u8. :D
But for input, you could just return `false` when you have an input that never ends.

Reply:
that is probably true, and yes we can do that indeed.

Or something like InputExtended can be something also.